### PR TITLE
Amsinh/W3C Tracing in SF SDK

### PR DESCRIPTION
**Pull request by ameyanator**

### CODE REVIEW

File: code.sln [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.27130.2027
+# Visual Studio Version 16
```
Comment (PavanKunapareddyMSFT) [inline code.sln:3]: why this change ?

File: nuprojs/Microsoft.ServiceFabric.Services.Remoting/Microsoft.ServiceFabric.Services.Remoting.nuproj [hunk 2]
```diff
@@ -98,5 +98,8 @@
       <Dependency Include="Microsoft.ServiceFabric.FabricTransport.Internal">
         <Version>[$(NugetPkg_Version_Microsoft_ServiceFabric_FabricTransport_Internal)]</Version>
       </Dependency>
+	  <Dependency Include="System.Diagnostics.DiagnosticSource">
+		<Version>[5.0.0.0]</Version>
```
Comment (PavanKunapareddyMSFT) [inline nuprojs/Microsoft.ServiceFabric.Services.Remoting/Microsoft.ServiceFabric.Services.Remoting.nuproj:102]: https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#version-ranges
Do not use exact version match. Also take dependency on lowest possible supported version

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        /// <returns><c>true</c> if this instance is present; otherwise, <c>false</c>.</returns>
+        internal static bool IsPresent()
+        {
+            return Activity.Current != null;
+        }
+
+        /// <summary>
+        /// Tries the get.
+        /// </summary>
+        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise.</returns>
```
Comment (PavanKunapareddyMSFT) [inline src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs:130]: Fix the documentation

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/FabricTransport/Runtime/FabricTransportServiceRemotingListener.cs [hunk 4]
```diff
@@ -40,18 +40,24 @@ public class FabricTransportServiceRemotingListener : IServiceRemotingListener
...
+        /// <param name="activitySourceName">Name of the ActivitySource to be used for tracing</param>
+        /// <param name="activitySourceVersion">Version of ActivitySource to be used for tracing</param>
         /// <param name="remotingListenerSettings">The settings for the listener</param>
         public FabricTransportServiceRemotingListener(
             ServiceContext serviceContext,
             IService serviceImplementation,
             FabricTransportRemotingListenerSettings remotingListenerSettings = null,
-            IServiceRemotingMessageSerializationProvider serializationProvider = null)
+            IServiceRemotingMessageSerializationProvider serializationProvider = null,
+            string activitySourceName = "Microsoft.ServiceFabric.ActivitySource",
```
Comment (PavanKunapareddyMSFT) [inline src/Microsoft.ServiceFabric.Services.Remoting/V2/FabricTransport/Runtime/FabricTransportServiceRemotingListener.cs:51]: We need not require this overload I guess. If user wants to override the activity source name, then they need to pass MessageDispatcher object with ServiceRemotingListener

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/Runtime/ServiceRemotingMessageDispatcher.cs [hunk 5]
```diff
@@ -39,11 +41,15 @@ public class ServiceRemotingMessageDispatcher : IServiceRemotingMessageHandler,
...
         /// <param name="serviceRemotingMessageBodyFactory">The factory that will be used by the dispatcher to create response message bodies.</param>
+        /// <param name="activitySourceName">Name of the ActivitySource to be used for tracing</param>
+        /// <param name="activitySourceVersion">Version of ActivitySource to be used for tracing</param>
         public ServiceRemotingMessageDispatcher(
             IEnumerable<Type> remotingTypes,
             ServiceContext serviceContext,
             object serviceImplementation,
-            IServiceRemotingMessageBodyFactory serviceRemotingMessageBodyFactory = null)
+            IServiceRemotingMessageBodyFactory serviceRemotingMessageBodyFactory = null,
+            string activitySourceName = "Microsoft.ServiceFabric.ActivitySource",
```
Comment (PavanKunapareddyMSFT) [inline src/Microsoft.ServiceFabric.Services.Remoting/V2/Runtime/ServiceRemotingMessageDispatcher.cs:51]: Microsoft.ServiceFabric.Services.Remoting.ActivitySource suits better as this is specific to service remoting calls only?

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+                remotingRequestRequestMessage.GetHeader().ActivityRequestId = currentActivity.Id.ToString();
+            }
+
+            foreach (var item in currentActivity.Baggage)
+            {
+                remotingRequestRequestMessage.GetHeader().ActivityIdBaggage.Add(item);
+            }
+        }
+
+        internal static void StartActivity(IServiceRemotingRequestMessage requestMessage, ActivitySource activitySource)
```
Comment (PavanKunapareddyMSFT) [inline src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs:60]: Change this to StartActivityIfNeeded or something similar

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+                parentId = requestMessage.GetHeader().ActivityRequestId;
+            }
+            else
+            {
+                // if activity not present
+                return;
+            }
+
+            // If there are any Listeners then create Activity using ActivitySource
+            var activity = activitySource.StartActivity("StatefulDatabaseIncomingRemoteCall", ActivityKind.Server, CreateActivityContextFromTraceParent(parentId));
```
Comment (PavanKunapareddyMSFT) [inline src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs:82]: StatefulDatabaseIncomingRemoteCall should be used as activity name

Comment (PavanKunapareddyMSFT) [inline src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs:82]: Does ActivitySource.CreateActivity also gives null activity if no listeners present? If yes, you dont need activitySourceListenerPresent  flag

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        internal static Activity Get()
+        {
+            return Activity.Current;
+        }
+
+        /// <summary>
+        /// Sets the specified activity.
+        /// </summary>
+        /// <param name="activity">The activity.</param>
+        internal static void Set(Activity activity)
```
Comment (PavanKunapareddyMSFT) [inline src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs:140]: Where are Set and Clear methods used?

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/Runtime/ServiceRemotingMessageDispatcher.cs [hunk 9]
```diff
@@ -67,13 +73,17 @@ public class ServiceRemotingMessageDispatcher : IServiceRemotingMessageHandler,
         /// <param name="serviceContext">Service context</param>
         /// <param name="serviceImplementation">Service implementation that implements interfaces of type <see cref="IService"/></param>
         /// <param name="serviceRemotingMessageBodyFactory">The factory that will be used by the dispatcher to create response message bodies.</param>
+        /// <param name="activitySourceName">Name of the ActivitySource to be used for tracing</param>
+        /// <param name="activitySourceVersion">Version of ActivitySource to be used for tracing</param>
         public ServiceRemotingMessageDispatcher(
```
Comment (PavanKunapareddyMSFT) [inline src/Microsoft.ServiceFabric.Services.Remoting/V2/Runtime/ServiceRemotingMessageDispatcher.cs:78]: I believe you need to create overloaded constructor for ActorDispatcher as well.

**Review by PavanKunapareddyMSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs (hunk 8 above)
Comment (ameyanator) [reply src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs:140]: Not used now, I was thinking if it would be used in the future. Removing them for now, will add later if needed for some other feature/implementation/scenario

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs (hunk 7 above)
Comment (ameyanator) [reply src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs:82]: The activity object itself is null. But then later once I start adding its attributes, I need to check later below if the ID is null or this flag. Both have same behavior

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs (hunk 3 above)
Comment (ameyanator) [reply src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs:130]: Made the change

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/Runtime/ServiceRemotingMessageDispatcher.cs (hunk 5 above)
Comment (ameyanator) [reply src/Microsoft.ServiceFabric.Services.Remoting/V2/Runtime/ServiceRemotingMessageDispatcher.cs:51]: Made the change

File: src/Microsoft.ServiceFabric.Services.Remoting/V2/Runtime/ServiceRemotingMessageDispatcher.cs (hunk 9 above)
Comment (ameyanator) [reply src/Microsoft.ServiceFabric.Services.Remoting/V2/Runtime/ServiceRemotingMessageDispatcher.cs:78]: Yep, Added the support

File: code.sln (hunk 1 above)
Comment (ameyanator) [reply code.sln:3]: I didn't make this change, it got pushed automatically in the code.sln file. Might be due to VS update maybe

File: nuprojs/Microsoft.ServiceFabric.Services.Remoting/Microsoft.ServiceFabric.Services.Remoting.nuproj (hunk 2 above)
Comment (ameyanator) [reply nuprojs/Microsoft.ServiceFabric.Services.Remoting/Microsoft.ServiceFabric.Services.Remoting.nuproj]: Made the change
